ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/txselection/BlockTransactionSelector.java

```diff
@@ -245,6 +265,69 @@ private void timeLimitedSelection() {
     }
   }
 
+  private void pluginTimeLimitedSelection(final long startTime) {
+    final CountDownLatch pluginSelectionDone = new CountDownLatch(1);
+
+    final var pluginTxSelectionTask =
+        new FutureTask<Void>(
+            () -> {
+              try {
+                LOG.debug(
+                    "Starting plugin transaction selection, run time capped at {}ms",
+                    pluginTxsSelectionMaxTime);
+                transactionSelectionService.selectPendingTransactions(
+                    this, blockSelectionContext.pendingBlockHeader());
+              } finally {
+                pluginSelectionDone.countDown();
+              }
+            },
+            null);
+
+    ethScheduler.scheduleBlockCreationTask(pluginTxSelectionTask);
+
+    try {
+      pluginTxSelectionTask.get(pluginTxsSelectionMaxTime, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException | ExecutionException e) {
+      if (isCancelled.get()) {
+        throw new CancellationException("Cancelled during plugin transaction selection");
+      }
+      LOG.warn("Error during block transaction selection", e);
+    } catch (TimeoutException e) {
+      // synchronize since we want to be sure that there is no concurrent state update
+      synchronized (isTimeout) {
```

Comment:
I wonder if synchronization is needed here as AtomicBoolean is is inherently thread-safe.